### ISSUE TITLE
[UX] Figure out a better way to show the "dirty form - needs save to apply changes" message so that it is within viewport.

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#1656]
I've been "playing around" with various issues related to either the Layout edit UI and the Views edit UI and #789 came to mind a lot, but there's another issue with how/where we show this message: there are cases where the message is out of the user's view, so the user has no idea that there is such a message shown (WTF situation).

I thought of auto-scrolling to the top of the page because this is what actually happens when for example the form is submitted and the "success!" message is shown. ...well, not auto-scrolling per se - it is a page load, but the effect is the same: message is shown where the user can see it. Anyways, this solution would be annoying and cause distraction to the users I think.

I also thought that we could use the newly-added feature of badges in the admin bar, but a) we can't be sure that the user has set the bar to be "sticky" at the top of the page, so no good because it will still be out of the viewport and b) we use badges for errors only AFAIK and the "dirty form" message is simply a warning. 

How about we make this message a yellow notification bar that floats at the top of the form instead? This way it's not outside the user's viewport where it can easily be missed. Note that if we go with this solution, we'll need to account for the case where the admin bar is set to be "sticky" at the top of the page and also the case of pages where we have "sticky" table headers at the top of the page too.

Other ideas?

---

**Description of the bug**
When I add a block to a layout, I click the "Add block" button in a region. I select the block I want from the block list, configure it, and add it to the page. Then, I have to SCROLL BACK TO the same place I was when I added the block, in order to drag it into the correct order. This can be very frustrating when you have lots of rows and many similarly named blocks. The page should not scroll at all when a block is added.

**Steps To Reproduce**
To reproduce the behavior:
1. Configure the home page layout at `admin/structure/layouts/manage/home`
2. Scroll to the footer, and click on `Add block`
3. Configure the block however, and click the `Add block` button
4. Can you see the footer region?

**Actual behavior**
The page has now scrolled to the top, away from the area I care about, to show me the message `This form has unsaved changes. Click "Save layout" to make changes permanent or "Cancel" to discard changes.`

**Expected behavior**
The page should not scroll away from what I care about to show me a message. 

**Related Issues**
https://github.com/backdrop/backdrop-issues/issues/2038

**Comment by klonos** [comment]
Mockups with the admin bar set to be sticky at the top of the page. Two versions: one with the message inside the admin bar:

![backdrop-dirty_form_message_and_save_cancel_buttons_sticky_inside_the_admin_bar](https://cloud.githubusercontent.com/assets/2423362/15541522/3d869616-22d0-11e6-92e9-bb23684c514c.gif)

...and one underneath it:

![backdrop-dirty_form_message_and_save_cancel_buttons_sticky_under_the_admin_bar](https://cloud.githubusercontent.com/assets/2423362/15541523/3d8d6d6a-22d0-11e6-8621-d06937ea0c75.gif)

**Comment by klonos** [comment]
...this would work great in the layouts edit form too.

**Comment by olafgrabienski** [comment]
> this would work great in the layouts edit form too.

Yes, and there is already a discussion about the Layout configuration page in https://github.com/backdrop/backdrop-issues/issues/2038, with the proposal to save changes on the Layout page by default (and to offer a respective *Cancel* button). That's an interesting idea but seems to have some challenges and may need some time.

I find the issue of the not-visible 'dirty form message' however quite urgent for the Layout page as it's a long page and I'm often not sure if I have still to save my changes. Usually I scroll to the top and check if there are changes but we can't expect that.

An easy fix would be to make the 'dirty form message' fixed via CSS. For the layout page it could be:

```
#layout-messages {
  position: fixed;
  z-index: 2;
}
```

I'd suggest to start with such an easy fix and to expand it then with existing proposals like https://github.com/backdrop/backdrop-issues/issues/789 and https://github.com/backdrop/backdrop-issues/issues/2038, https://github.com/backdrop/backdrop-issues/issues/2305, and https://github.com/backdrop/backdrop-issues/issues/2430.

**Comment by klonos** [comment]
...as I have also mentioned in #789, this might become a non-issue if the current implementation of #566 gets accepted:

![backdrop-issue1656](https://user-images.githubusercontent.com/2423362/42351176-0030778c-80f7-11e8-810d-890c20567e75.gif)

PR: https://github.com/backdrop/backdrop/pull/2237

**Comment by olafgrabienski** [comment]
Recently I've worked a lot on Layout block configuration pages and several times I wasn't sure if I have to save the configuration due to a potential out-of-viewport warning message.

While I find #566 is still a good idea (got lost somehow, we should revive it), I'd propose to take the famous baby steps, making the warning message on the Layout configuration page stick to the top. Here's a proposal:

```
#layout-messages .warning {
  position: fixed;
  left: 0;
  top: 4.25em;
  opacity: 0.9;
  z-index: 2;
}
@media only screen and (min-width: 48em) {
  #layout-messages .warning {
  top: 7.25em;
}
@media only screen and (min-width: 62em) {
  #layout-messages .warning {
  left: unset;
  top: 8.75em;
}
```

My suggestion to not declare `top: 0` but to leave some space, so that important elements of the page remain readable and touchable / clickable without the need of scrolling, e.g. the title and the breadcrumb. The opacity change takes situations into account where -- in spite of the extra space -- some elements are partially covered by the message.

I could make a PR but would love to get some feedback before (below are some screenshots). Also, I'm not sure where to make the changes, probably in `core/modules/system/css/messages.theme.css`.

Screenshots
------------

**Notebook:**

![screen-backdrop-issue-1656-proposal-notebook](https://user-images.githubusercontent.com/14188307/65246090-2afff580-daee-11e9-8866-4c9213d53d9c.png)

**Tablet:**

![screen-backdrop-issue-1656-proposal-tablet](https://user-images.githubusercontent.com/14188307/65246092-2afff580-daee-11e9-9424-df09aa8300b2.png)


**Phone:**

![screen-backdrop-issue-1656-proposal-phone](https://user-images.githubusercontent.com/14188307/65246425-db6df980-daee-11e9-9040-8a7642fb2571.png)

**Comment by jenlampton** [comment]
oops, I just created [the exact same issue](https://github.com/backdrop/backdrop-issues/issues/4074)!  Im adding some keywords here so I can find this next time I (or someone else) searches for it :) 

layout page scrolls away from block after block is added
